### ISSUE TITLE
Bug: `AsyncExitStack` with default typevar check failed when used in union type

### ISSUE THREAD

**Issue by yanyongyu** [microsoft/pyright#7764]
**Describe the bug**

Typeshed changed the signature of exit stack and context manager in [commit](https://github.com/python/typeshed/commit/f274c78fe51dc9013004fcbf437631ecf6a4fae9). The return type of the cm protocol becomes a typevar (with default type). It seems pyright failed to use the default type of the typevar when using exit stack in union type (such as `AsyncExitStack | None`). The detailed error can be found in the playground.

**Code or Screenshots**

[Playground](https://pyright-play.net/?code=GYJw9gtgBAxmB2AXApgD0QGwJYCMpYgAcwREoBBAZwE94YBRVLRAZUQEMYBrAKFEiiJqhZJXxESZACohOyHJy5ThyPuGhDCWeAHNxxUlBbIMwADRQQyAG7J2GAPqbVPHjAztKY8gC4eUAKhPWhgoABNkYCgHB3ZkJGQQGIAKShNgAEooAFoAPiN0v0Diy2REAFcQeCg0039A%2BoDgunDI6Ni0ZhTGktrzHuK0GCcVH0EVAG0AIU9kRhhkQkQsBABdKAAfKAA5BGQzAcChh2t7cuQxmbT5xeWETZ29g5LAxFkFhW4xmTlPpRUHrt4Pselk8o9gUUXlBCJ5KK4eM1QhEohB2NpkhkoU0oABeCiY1zFSgcL4UGh0RjMNiKQF7PHkkJU1ikriE4rsADu6LIJMUADp4igkkiHHAEuhkuwMkTAnzuA4xlQmUwWbT8crKaqadx2YEuTyaqyHIKEiKKcNxShJdKeEA)

```python
from contextlib import AsyncExitStack
from types import TracebackType
from typing import Self, reveal_type


class A:
    async def __aenter__(self) -> Self:
        return self
    
    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_value: BaseException | None,
        traceback: TracebackType | None,
    ) -> None:
        pass


async def main():
    a = A()

    stack: AsyncExitStack | None = AsyncExitStack()
    await stack.enter_async_context(a)  # error

    stack_: AsyncExitStack = AsyncExitStack()
    await stack_.enter_async_context(a)  # ok

```

This also occurs when using `ExitStack`.

**VS Code extension or command-line**

error with command line 1.1.360, ok with pylance stable 2024.4.1 (pyright 1.1.356, typshed not updated)

**Comment by erictraut** [comment]
Thanks for the bug report. I agree this isn't working as it should in the bidirectional type inference case. If you remove the type declaration for `stack`, it works correctly (`stack = AsyncExitStack()`).

**Comment by yanyongyu** [comment]
Removing the type declaration works, but in my use case the stack is optional and initialized with `None`. so this error occurs when i declare the type of stack with `AsyncExitStack | None`. :joy:

**Comment by erictraut** [comment]
This issue specifically affects generic classes that do not define either a `__new__` or `__init__` method _and_ bidirectional type inference is used for the constructor call _and_ the expected type is a union.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.361, which I just published.
